crates/bitwarden/src/tool/generators/passphrase.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{error::Result, wordlist::EFF_LONG_WORD_LIST};
+use rand::{seq::SliceRandom, Rng, RngCore};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Passphrase generator request.
+///
+/// The default separator is `-` and default number of words is 3.
+#[derive(Serialize, Deserialize, Debug, JsonSchema, Default)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct PassphraseGeneratorRequest {
+    pub num_words: Option<u8>,
+    pub word_separator: Option<String>,
+    pub capitalize: Option<bool>,
+    pub include_number: Option<bool>,
+}
+
+const DEFAULT_PASSPHRASE_NUM_WORDS: u8 = 3;
+const DEFAULT_PASSPHRASE_SEPARATOR: char = ' ';
+
+pub(super) fn passphrase(input: PassphraseGeneratorRequest) -> Result<String> {
+    passphrase_with_rng(rand::thread_rng(), input)
+}
+
+fn passphrase_with_rng(mut rng: impl RngCore, input: PassphraseGeneratorRequest) -> Result<String> {
```

Comment:
I've added some checks for the number of words that we didn't have before, so I guess this is technically solved 😄. I'm returning an error instead of clamping the number of words to the 3..20 range like the TypeScript code does. This to me seems more appropriate for an SDK, as there's no feedback to the user if we were to clamp the value. The UI should be the one limiting the range to the users, what do you think?

Reply:
I'm interpreting your comment as "The SDK shouldn't control the user experience", and I agree that doesn't make much sense. That said, I'd wonder two things.

1. Should there be a wrapping library that *does* control the user experience. That is, should there be a `bitwarden_services` library or similar that is responsible for user-level validation?
2. Should the SDK limit the number of words at all? (This one's rhetorical--it shouldn't, since that would apply an arbitrary limit to the user experience.)

Reply:
To me it's not so much about not controlling the user experience, but more about just rejecting invalid options instead of silently changing them under the user to make them valid.

If a SDK user requests a 40 word passphrase, we have two ways of going about it that I consider correct:
- Either not limiting the option at all, and returning a 40 word passphrase.
- Or erroring out and mentioning we limit the generation to 20 characters.

On the other hand, disregarding the user's request for a 40 word passphrase, silently changing the value to 20 and returning a 20 word passphrase is just incorrect to me.

I think we should enforce the limits at the SDK level, mostly to avoid possible inconsistencies in all the clients using the SDK. 
At the same time, the UI should do validation to make sure that users can't select an invalid combination of options, if that can be exposed as a generic library for all the clients that'd be great

Reply:
> I think we should enforce the limits at the SDK level, mostly to avoid possible inconsistencies in all the clients using the SDK.

If there are technical limitations to generating passphrases, then setting SDK bounds is completely cool. I would not, however, do so arbitrarily in the generator. This is a policy decision, and should be applied by the SDK's caller. As to why? Consider what happens if circumstances change to make >20-word passphrases reasonable. You'll need to re-release the SDK and all downstream products to reverse this change.

That said, I'm not entirely against this, so long as the generator isn't enforcing the policy. For example, you can introduce a "policy free" struct that is used by this method, implement a conversion trait, and enforce the policy in the trait's impl. That way, the generator doesn't need to change when the policy does.

I've created an example of how this can work [on the rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=76aa3acd3a4bbfafa8b5ca32167364d0) using `TryFrom<...>`. While I'm not enthused about keeping the policy compiled in the SDK, using the trait-based solution introduces the flexibility required to migrate policy if needed.

Reply:
> Consider what happens if circumstances change to make >20-word passphrases reasonable. You'll need to re-release the SDK and all downstream products to reverse this change.

The current architecture is centered around the reverse. We want to enforce consistency across clients, one way to do this is by moving these concerns into the SDK. A future iteration could be to explore if we should expose available UI options so that updating the SDK would also change the presentation layer.

Reply:
The 20-word limitation is built in to the SDK. This is about making sure we're using rust types to implement the default so that we can easily extend it later.

If this were C# code, it would be akin to introducing an interface for mocking purposes.

Reply:
I view the validation inside the generate function to be typical out of bounds checks. Consider it part of the generators specification, it can generate words between 3 and 20 character, it needs a valid separator.

I would expect C# to behave similarly, I should not be able to pass in wrong values without getting back an error.

I do think we can remove the optional aspects from the request, since the UI constraints will always have options for this. I have a commit for this, if you can verify it satisfies the concern.

Reply:
[Lines 50-52](https://github.com/bitwarden/sdk/blob/16207a6127133e70f37ec247ea79a2aaa5472959/crates/bitwarden/src/tool/generators/passphrase.rs#L50-L52) are the problem. If there's a conflict with policy set in the administrative console, they should be ignored. 

Using `TryFrom<T>` and an intermediate options type seems like the proper solution from multiple perspectives:

* It does everything you need it to do without changing callsites.
* It provides a wide path forward for policy overrides.
* It isolates the decision making code as you've requested elsewhere in the PR.
* It reduces the scope of unit tests

Reply:
A policy cannot and will not be a allowed to override the generators inherit constraints. The generator specification is explicit in that it can generate a passphrase between 3 and 20 words, it can capitalize, include numbers and supports different separators.

Passing in a policy that says `min_words = 21` will always fail, either the policy validation will ask the user to change their settings, or fail due to the generator attempting to generate to long passphrase.

As for the `TryFrom` pattern, I don't believe it's a good fit here for several reasons.

1. We will not allow consumers to implement the `TryFrom` themselves. Doing so would let them violate the policies or the generators built in constraints.
2. We would need a wrapper struct `PolicyOptions(options, Some(policy))` just to handle `TryFrom` with multiple values.

We do not yet know how policies will be represented which is why I dislike "future proofing" code. I would expect policies to each be their own struct which can implement functions for validation, which would be a layer before the generators own validations.

Reply:
> A policy cannot and will not be a allowed to override the generators inherit constraints.

The generator could output gigabytes of output, given enough memory. Say that every word is 1kb in size. 20 words is 20kb. This isn't an inherent constraint, it's arbitrary. And it's an arbitrary constraint that *today* is overridable in the admin console, and that I confirmed must continue to work.

> We will not allow consumers to implement the TryFrom themselves.

Consumers within the SDK boundary or outside of it? Because this doesn't change the FFI at all. And if it's consumers within the SDK, then this argument is silly. Anyone on the tools team could come in and make this change tomorrow. In fact, the reason I want to introduce indirection here--whether using `TryFrom<T>` or some other code that runs before the method is called--is so that I can implement the policy override.

> I would expect policies to each be their own struct which can implement functions for validation, which would be a layer called by the generator, either before or after performing it's own boundary validation.

Then I expect the exact same thing from your code.

This whole conversation stemmed from my desire to have an *infallible* generator. Dani wants to check the arguments, and I gave him a way to do so. We're now stuck going back and forth about the mechanism. I don't care about that, it just solved the problem elegantly.

-----

I care that:

1. The generator function does not check policy.
2. That the arbitrary limits applied to the consumer product can be overridden by administrators.

`TryFrom<T>` is unacceptable, so let's move on. Choose your own design that accomplishes the following:

❌ Pass a type that requires no validation into the generator.
❌ Do not pass the FFI type into the generator.
❌ Write the generator to be infallible and/or only fail by panicking.